subsys/power/power.c

```diff
@@ -101,16 +101,17 @@ enum power_states _sys_suspend(s32_t ticks)
 	post_ops_done = 0;
 	sys_pm_notify_power_state_entry(pm_state);
 
-	if (deep_sleep) {
 #if CONFIG_DEVICE_POWER_MANAGEMENT
-		/* Suspend peripherals. */
-		if (sys_pm_suspend_devices()) {
-			LOG_ERR("System level device suspend failed!");
-			sys_pm_notify_power_state_exit(pm_state);
-			pm_state = SYS_POWER_STATE_ACTIVE;
-			return pm_state;
-		}
+	/* Suspend peripherals. */
+	if (sys_pm_suspend_devices(ticks)) {
```

Comment:
This is the part we should really think about IMHO.
At the moment we are coupling device power state with the system state, but I am not sure if we should do that:
- The devices which are already not used should be suspended earlier to save the power.
- The devices might need time to suspend/resume, and this part blocks whole system until this is done.
- We should not resume devices which are not used as this will increase power consumption.

Reply:
> This is the part we should really think about IMHO.
> At the moment we are coupling device power state with the system state, but I am not sure if we should do that:
> 
> * The devices which are already not used should be suspended earlier to save the power.
> * The devices might need time to suspend/resume, and this part blocks whole system until this is done.
> * We should not resume devices which are not used as this will increase power consumption.

yes, but that should be done by device driver, peripheral should decide suspend/resume by itself, that will be triggered by himself's logic(for example, there is no traffic for wifi chipset connected to AP/MCU, wifi chipset will go to suspend state). This PR is that system decide to suspend, sys_pm_suspend_devices is to notify peripherals, at least peripheral know system's state, if it want to work, wakeup cpu first. Further peripheral can enter low power state according system state or system suspend ticks. Yes, it may take some time and block the whole system, but we have to do that(linux do the notify also one by one), let end users to tuning system suspend ticks and the time of devices entering lower state. 
And currently we let devices enter suspend state only when system going to deep sleep, but what about system in sleep state(not deep sleep)? I think we should let peripherals know system state at least.
I'm not familiar with the design of Zephyr's power management, maybe there is something wrong with what I said. @pizi-nordic

Reply:
> This is the part we should really think about IMHO.
> At the moment we are coupling device power state with the system state, but I am not sure if we should do that:
> 
>     * The devices which are already not used should be suspended earlier to save the power.
> 
>     * The devices might need time to suspend/resume, and this part blocks whole system until this is done.
> 
>     * We should not resume devices which are not used as this will increase power consumption.

Reviewed power management code more, I think my patch doesn't follow Zephyr power management infrastructure. But I still think we should consider device power management even when system going to sleep state(not deep sleep) in order to save more power if only enable central method device power management.  
```
Central method
==============

In this method device power management is mostly done inside
:code:`sys_suspend()` along with entering a CPU or SOC power state.

If a decision to enter deep sleep is made, the implementation would enter it
only after checking if the devices are not in the middle of a hardware
transaction that cannot be interrupted. This method can be used in
implementations where the applications and components using devices are not
expected to be power aware and do not implement device power management.
```
My basic idea is to let device enter DEVICE_PM_LOW_POWER_STATE state in that case, and resume when system also resume to active state, also will add basic policy function for this logic, end user can adjust it based on the SOC they use. And this logic will be under marco !CONFIG_DEVICE_IDLE_PM. 

Really appreciate for your comments. @pizi-nordic